src/sentry/api/endpoints/organization_traces.py

```diff
@@ -357,8 +379,261 @@ def execute(self, offset: int, limit: int):
+        # based on some heuristics for that type of trace
+        traces_fallback_info: MutableMapping[str, tuple[str, str, float]] = {}
+
+        # This is the name of the first span in the trace that will be used if
+        # no other candidates names are found
+        traces_default_info: MutableMapping[str, tuple[str, str, float]] = {}
+
+        # Normally, the name given to a trace is the name of the first root transaction
+        # found within the trace.
+        #
+        # But there are some cases where traces do not have any root transactions. For
+        # these traces, we try to pick out a name from the first span that is a good
+        # candidate for the trace name.
+        for span in spans:
+            if span["trace"] in traces_primary_info:
+                continue
+
+            name: tuple[str, str, float] = (
+                span["project"],
+                span["transaction"],
+                # to minmimize the impact of floating point errors,
+                # multiply by 1e3 first then do the subtraction
+                # once we move to eap_items, this can be just `span["span.duration"]`
+                span["precise.finish_ts"] * 1e3 - span["precise.start_ts"] * 1e3,
+            )
+
+            # The underlying column is a Nullable(UInt64) but we write a default of 0 to it.
+            # So make sure to handle both in case something changes.
+            if not span["parent_span"] or int(span["parent_span"], 16) == 0:
+                traces_primary_info[span["trace"]] = name
+
+            if span["trace"] in traces_fallback_info:
+                continue
+
+            # This span is a good candidate for the trace name so use it.
+            if span["trace"] not in traces_fallback_info and is_trace_name_candidate(span):
+                traces_fallback_info[span["trace"]] = name
+
+            if span["trace"] in traces_default_info:
+                continue
+
+            # This is the first span in this trace.
+            traces_default_info[span["trace"]] = name
+
+        def get_trace_info(trace: str) -> tuple[str, str, float] | tuple[None, None, None]:
+            if trace in traces_primary_info:
+                return traces_primary_info[trace]
+
+            if trace in traces_fallback_info:
+                return traces_fallback_info[trace]
+
+            if trace in traces_default_info:
+                return traces_default_info[trace]
+
+            return (None, None, None)
+
+        for trace in traces:
+            info = get_trace_info(trace["trace"])
+            trace["project"] = info[0]
+            trace["name"] = info[1]
+            trace["rootDuration"] = info[2]
+
+            trace["numErrors"] = traces_errors.get(trace["trace"], 0)
+            trace["numOccurrences"] = traces_occurrences.get(trace["trace"], 0)
+            trace["breakdowns"] = traces_breakdowns[trace["trace"]]
+
+    def get_traces_rpc(self, projects: list[Project]):
```

Comment:
What are your thoughts on moving this to spans_rpc.py with the other rpc calls? 🤔
Might just be a missplaced purism thing on my part where i'm trying to keep all this proto code out of the endpoints.

Reply:
going to keep this here for now as this does some finer tweaking of the proto that doesn't fit well in the generic `spans_rpc.py` file.

For example, it overrides the projects in the snuba params and generates an `IN` condition for the original list of projects.